src/main/scala/firrtl/DiGraph.scala

```diff
@@ -0,0 +1,158 @@
+package firrtl
+
+import scala.collection.immutable.{HashSet, HashMap}
+import scala.collection.mutable
+import scala.collection.mutable.MultiMap
+
+class MutableDiGraph[T](
+  val edgeData: MultiMap[T,T] = new mutable.HashMap[T, mutable.Set[T]] with MultiMap[T, T]) {
+  def contains(v: T) = edgeData.contains(v)
+  def getVertices = edgeData.keys
+  def getEdges(v: T) = edgeData(v)
+  def addVertex(v: T): T = {
+    edgeData.getOrElseUpdate(v,new mutable.HashSet[T])
+    v
+  }
+  // Add v to keys to maintain invariant
+  def addEdge(u: T, v: T) = {
+    edgeData.getOrElseUpdate(v, new mutable.HashSet[T])
+    edgeData.addBinding(u,v)
+  }
+}
+
+object DiGraph {
+  def apply[T](mdg: MutableDiGraph[T]) = new DiGraph((mdg.edgeData mapValues { _.toSet }).toMap[T, Set[T]])
+  def apply[T](edgeData: MultiMap[T,T]) = new DiGraph((edgeData mapValues { _.toSet }).toMap[T, Set[T]])
+}
+
+class DiGraph[T] (val edges: Map[T, Set[T]]) {
+
+  def getVertices = edges.keys
+  def getEdges(v: T) = edges.getOrElse(v, new HashSet[T])
+
+  // Graph must be acyclic for valid linearization
+  def linearize(root: T) = {
+    val order = new mutable.ArrayBuffer[T]
+    val visited = new mutable.HashSet[T]
+    def explore(v: T): Unit = {
+      visited += v
+      for (u  <- getEdges(v)) {
+        if (!visited.contains(u)) {
+          explore(u)
+        }
+      }
+      order.append(v)
+    }
+    explore(root)
+    order.reverse.toList
+  }
+
+  def doBFS(root: T) = {
+    val prev = new mutable.HashMap[T,T]
+    val queue = new mutable.Queue[T]
+    queue.enqueue(root)
+    while (!queue.isEmpty) {
+      val u = queue.dequeue
+      for (v <- getEdges(u)) {
+        if (!prev.contains(v)) {
+          prev(v) = u
+          queue.enqueue(v)
+        }
+      }
+    }
+    prev
+  }
+
+  def reachabilityBFS(root: T) = doBFS(root).keys.toSet
+
+  def path(start: T, end: T) = {
+    val nodePath = new mutable.ArrayBuffer[T]
+    val prev = doBFS(start)
+    nodePath += end
+    while (nodePath.last != start) {
+      nodePath += prev(nodePath.last)
+    }
+    nodePath.toList.reverse
+  }
+
+  def findSCCs = {
```

Comment:
Strongly Connected Component. I more or less try to use standard graph jargon throughout to keep it as normal as possible.